### PR TITLE
chore: Bump go.mod to use go 1.23

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes #N/A 

**Description**

This updates Karpenter Provider to use the latest version of go, `go1.23`

**How was this change tested?**

`make presubmit`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: charts/karpenter-crd/templates/karpenter.k8s.aws_ec2nodeclasses.yaml [hunk 1]
```diff
@@ -940,6 +913,9 @@ spec:
                           RootVolume is a flag indicating if this device is mounted as kubelet root dir. You can
                           configure at most one root volume in BlockDeviceMappings.
                         type: boolean
+                    required:
+                      - deviceName
+                      - ebs
```
Comment (jmdeal) [inline charts/karpenter-crd/templates/karpenter.k8s.aws_ec2nodeclasses.yaml:918]: I know this came from the new kubebuilder version, but this stands out as a bit of a red flag to me. Have we verified what the consequences of this change are?

Comment (jonathan-innis) [reply charts/karpenter-crd/templates/karpenter.k8s.aws_ec2nodeclasses.yaml]: Good catch -- updated so we don't change the fields that are required here -- though the API may be invalid anyways without those fields being required so we may want to validate that in a subsequent PR

**Review by jonathan-innis — COMMENTED** [review]
/karpenter snapshot

**Review by jmdeal — APPROVED** [review]
LGTM 🚀

**Pull request merged**